### PR TITLE
[pytorch] [trcomp] Add support for PT 1.12 SM Training Compiler

### PR DESCRIPTION
**Pull request by Vatshank**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Adds support for SageMaker Training Compiler to PyTorch 1.12.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [x] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+RUN pip uninstall -y sagemaker-training \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-training>=4.2.6"
+
+# Install SageMaker PyTorch Training Toolkit
+RUN pip uninstall -y sagemaker-pytorch-training \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-pytorch-training>=2.6.1,<3"
+
+# CVE Fixes
+RUN pip install --no-cache-dir -U "protobuf<3.20.0"
+
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:103]: Missing license https://github.com/aws/deep-learning-containers/blob/master/pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu#L221

Comment (harishneit) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:103]: we already have this file on the filesystem as we inherit from PT 1.12 container right?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 2]
```diff
@@ -633,6 +633,10 @@ def test_pip_check(image):
 
     :param image: ECR image URI
     """
+
+    if "pytorch" in image and "trcomp" in image:
```
Comment (tejaschumbalkar) [inline test/dlc_tests/sanity/test_pre_release.py:637]: why is this skipped? Do you mean the pip check will always be broken in the image?

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 1 above)
Comment (tejaschumbalkar) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:117]: do you need a new one for this image explicitly? how is this done for hf image?

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 3]
```diff
@@ -136,6 +136,9 @@ def test_pytorch_train_dgl_cpu(pytorch_training, ec2_connection, cpu_only, py3_o
 @pytest.mark.model("mnist")
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_GPU_INSTANCE_TYPE, indirect=True)
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:137]: All the pytest skips in this file are applicable for huggingface images as well?

File: test/dlc_tests/ec2/test_smdebug.py [hunk 4]
```diff
@@ -4,7 +4,10 @@
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/test_smdebug.py:4]: Same here, does this apply for hf images as well?

Comment (harishneit) [reply test/dlc_tests/ec2/test_smdebug.py:4]: yes, we don't support horovod with trcomp for pytorch

File: test/dlc_tests/sanity/test_pre_release.py (hunk 2 above)
Comment (harishneit) [reply test/dlc_tests/sanity/test_pre_release.py:637]: yes. pytorch xla lists tpu-client as a runtime dependency but it's not installed.

**Comment by harishneit** [comment]
> The `version` in `pytorch/buildspec-trcomp.yml` needs to be changes to `1.12.1` since we are using the latest pytorch image that has v1.12.1

pytorch-xla only has major releases, we can only do 1.12.0

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+# Install TorchAudio
+RUN pip uninstall -y torchaudio \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall ${TORCHAUDIO_URL} \
+ && rm -rf ${TORCHAUDIO_URL}
+
+ARG CONDA_PREFIX=/opt/conda
+
+# Numpy version gets downgraded and becomes incompatible with PyTorch wheel in the above conda command.
+RUN pip install numpy --upgrade
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:82]: lets remove this, numpy is already installed using conda. See #2416 for the context on incompatibility

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+# Fix library links
+RUN ln -s ${CONDA_PREFIX}/lib/libmkl_intel_lp64.so ${CONDA_PREFIX}/lib/libmkl_intel_lp64.so.1 \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:85]: isn't conda linked to blas runtime provider? Are we explicitly linking it to mkl?

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# Install SageMaker Training Toolkit
+RUN pip uninstall -y sagemaker-training \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-training>=4.2.6"
+
+# Install SageMaker PyTorch Training Toolkit
+RUN pip uninstall -y sagemaker-pytorch-training \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-pytorch-training>=2.6.1,<3"
+
+# CVE Fixes
+RUN pip install --no-cache-dir -U "protobuf<3.20.0"
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:102]: is this required since the base image is updated?

Comment (harishneit) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:102]: yes, pt-xla doesn't update protobuf

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 5 above)
Comment (tejaschumbalkar) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:82]: which one?

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 7 above)
Comment (tejaschumbalkar) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:102]: the protobuf from base image is `protobuf==3.19.6` Does the pytorch binary installation removes this package? this command does not have a `--upgrade` flag so i believe if the package is installed already, this command will have no effect

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 6 above)
Comment (harishneit) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:85]: pytorch xla requires mkl to be in the ld_library_path

File: dlc_developer_config.toml [hunk 8]
```diff
@@ -20,7 +20,7 @@ huggingface_trcomp_mode = false
 # Please only set it to True if you are preparing a TRCOMP related PR
 # Do remember to revert it back to False before merging any PR (including TRCOMP dedicated PR)
 # This mode is used to build PT1.12 and above DLC
-trcomp_mode = false
+trcomp_mode = true
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:23]: Revert the changes  in this file to prepare the PR for merge

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+ && ln -s ${CONDA_PREFIX}/lib/libmkl_intel_thread.so ${CONDA_PREFIX}/lib/libmkl_intel_thread.so.1 \
+ && ln -s ${CONDA_PREFIX}/lib/libmkl_core.so ${CONDA_PREFIX}/lib/libmkl_core.so.1 \
+ && ln -s ${CONDA_PREFIX}/lib/libmkl_gnu_thread.so ${CONDA_PREFIX}/lib/libmkl_gnu_thread.so.1
+
+# Install SM Distributed DataParallel binary
+RUN SMDATAPARALLEL_PT=1 pip install --no-cache-dir ${SMD_DATA_PARALLEL_URL} --force-reinstall
+
+# Install SageMaker Training Toolkit
+RUN pip uninstall -y sagemaker-training \
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-training>=4.2.6"
```
Comment (Lokiiiiii) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:95]:
```suggestion
 && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-training>=4.3.2"
```

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 10]
```diff
@@ -258,8 +261,11 @@ def test_pytorch_amp(pytorch_training, ec2_connection, gpu_only, ec2_instance_ty
 @pytest.mark.integration("pt_s3_plugin_gpu")
 @pytest.mark.model("N/A")
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_GPU_INSTANCE_TYPE, indirect=True)
-def test_pytorch_s3_plugin_gpu(pytorch_training, ec2_connection, ec2_instance_type, outside_versions_skip):
+def test_pytorch_s3_plugin_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type, outside_versions_skip):
     outside_versions_skip(pytorch_training, "1.8.0", "1.12.1")
+    _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
+    if 'trcomp' in pytorch_training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+        pytest.skip(f"Image {pytorch_training} doesn't s3. Hence test is skipped.")
```
Comment (Lokiiiiii) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:268]: Fix Grammar

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+ && pip install --no-deps --no-cache-dir -U --force-reinstall "sagemaker-pytorch-training>=2.6.1,<3"
+
+# CVE Fixes
+RUN pip install --no-cache-dir -U "protobuf<3.20.0"
+RUN apt-get update -y && \
+    apt-get upgrade -y openssl && \
+    apt-get clean
+
+# Uninstall Torch Data
+RUN pip uninstall -y torchdata
```
Comment (Lokiiiiii) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:108]: Add an explanation for why we uninstall torchdata

File: test/dlc_tests/ec2/test_smdebug.py [hunk 12]
```diff
@@ -25,6 +28,10 @@
 def test_smdebug_gpu(training, ec2_connection, region, ec2_instance_type, gpu_only, py3_only):
     if test_utils.is_image_incompatible_with_instance_type(training, ec2_instance_type):
         pytest.skip(f"Image {training} is incompatible with instance type {ec2_instance_type}")
+
+    _, image_framework_version = get_framework_and_version_from_tag(training)
+    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+        pytest.skip(f"Image {training} doesn't s3. Hence test is skipped.")
```
Comment (Lokiiiiii) [inline test/dlc_tests/ec2/test_smdebug.py:34]: Fix Grammar

File: test/dlc_tests/ec2/test_smdebug.py [hunk 13]
```diff
@@ -56,6 +63,9 @@ def test_smprofiler_gpu(
     # This code needs to be modified past reInvent 2020
     if test_utils.is_image_incompatible_with_instance_type(training, ec2_instance_type):
         pytest.skip(f"Image {training} is incompatible with instance type {ec2_instance_type}")
+    _, image_framework_version = get_framework_and_version_from_tag(training)
+    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+        pytest.skip(f"Image {training} doesn't s3. Hence test is skipped.")
```
Comment (Lokiiiiii) [inline test/dlc_tests/ec2/test_smdebug.py:68]: Fix Grammar

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: dlc_developer_config.toml [hunk 14]
```diff
@@ -53,7 +53,7 @@ sagemaker_local_tests = false
 # "standard" --> run standard sagemaker remote tests from test/sagemaker_tests
 # "rc" --> run release_candidate_integration tests
 # "efa" --> run efa sagemaker tests
-sagemaker_remote_tests = "off"
+sagemaker_remote_tests = "standard"
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:56]: revert the file changes

**Comment by tejaschumbalkar** [comment]
Please update PR description/title. Make the PR ready for review

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 9 above)
Comment (Vatshank) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu]: Updated

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py (hunk 10 above)
Comment (Vatshank) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py]: Done

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu (hunk 11 above)
Comment (Vatshank) [reply pytorch/training/docker/1.12/py3/cu113/Dockerfile.trcomp.gpu:108]: Done

File: test/dlc_tests/ec2/test_smdebug.py (hunk 12 above)
Comment (Vatshank) [reply test/dlc_tests/ec2/test_smdebug.py]: Done

File: test/dlc_tests/ec2/test_smdebug.py (hunk 13 above)
Comment (Vatshank) [reply test/dlc_tests/ec2/test_smdebug.py]: Done

File: dlc_developer_config.toml (hunk 14 above)
Comment (Vatshank) [reply dlc_developer_config.toml]: Reverted

**Review by tejaschumbalkar — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]
